Implement live preview for site.json

jonahtanjz opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [x] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Resolves #48.

A follow up to #1514, which supported live preview for the pages attribute. This PR will support live preview for all the other attributes in `site.json`. A server restart is no longer required for these changes to be reflected when serving the site through `markbind serve`. Since most of the attributes in `site.json` (except `baseUrl`, `pages` and `deploy`) will affect majority of the pages, hence all pages will be regenerated when any of these attributes (except those listed above) is modified. 

For the `style` attribute, if a `bootswatch` theme is specified/added, its `bootstrap.min.css` theme file will overwrite the existing file and all pages are rebuilt. Similarly, if a `bootswatch` theme is removed, the default `bootstrap.min.css`, which is located at `core-web/asset/css` will overwrite the existing file.

**Anything you'd like to highlight / discuss:**
Would appreciate if you can let me know if I missed any attributes or if any of the attributes (other than those listed above) that does not require all pages to be rebuilt when it is modified.

**Testing instructions:**
1. Serve a MarkBind site using `markbind serve`
2. Open and edit the any of the attributes in site.json
3. Live preview should be updated to reflect the latest changes
<!--
  Any special testing instructions in not including our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Implement live preview for site.json

When attributes other than pages in site.json are modified, a server 
restart is required to reflect the changes on the live preview. This can 
become a hassle to the user as they will need to frequently restart 
the server when trying out different site configuration.

Let's support live preview for all the attributes in site.json so that the 
latest changes can be reflected without restarting the server.
<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!--
  Prefix your pull request title with [WIP] if any of these are not complete yet

  Tick non-applicable items as well
-->

- [x] Updated the documentation for feature additions and enhancements
- [x] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No blatantly unrelated changes
- [x] Pinged someone for a review!

raysonkoh commented on packages/core/src/Site/index.js:
```diff
@@ -889,6 +919,35 @@ class Site {
+  async handleIgnoreReload(oldIgnore) {
+    const assetsToRemoved = _.difference(this.siteConfig.ignore, oldIgnore);
+
+    if (!_.isEqual(oldIgnore, this.siteConfig.ignore)) {
```
Hmm, is this check neccessary? Would there be a case where `assetsToRemoved` is truthy but `oldIgnore == this.siteConfig.ignore`?

jonahtanjz replied:
I believe this check is necessary. For example, if one or more entries are removed from `ignore`, then `assetsToRemove` will be empty but `oldIgnore !== this.siteConfig.ignore` so the assets will be rebuilt. Another scenario is when `oldIgnore == this.siteConfig.ignore`, then in this case, if they are equal then no assets will need to be built or removed.

raysonkoh replied:
Ah okay got it. We can leave this in then 👍

raysonkoh commented on packages/core/src/Site/index.js:
```diff
@@ -889,6 +919,35 @@ class Site {
+      if (this.siteConfig.style.bootstrapTheme) {
+        await this.copyBootswatchTheme();
+      } else {
+        const defaultBootstrapTheme = require.resolve('@markbind/core-web/asset/css/bootstrap.min.css');
```
It seems like this block of logic is similar to `Site#copyBootswatchTheme()`. Would it be better to modify `copyBootswatchTheme()` to also account for copying the default bootstrap theme and then using calling that function in here instead?

jonahtanjz replied:
Sounds good. I have updated the method to copy the default theme if applicable during live reload of site.json.

raysonkoh reviewed (commented):
Neat code! Would it be possible to add some tests for this?

raysonkoh commented on packages/core/src/Site/index.js:
```diff
@@ -1282,15 +1275,18 @@ class Site {
+  copyBootstrapTheme(toCopyDefault) {
     const { theme } = this.siteConfig;
-    if (!theme || !_.has(SUPPORTED_THEMES_PATHS, theme)) {
+    if (!toCopyDefault && (!theme || !_.has(SUPPORTED_THEMES_PATHS, theme))) {
```
Minor nit: would it be possible to remove the need for the `toCopyDefault`? It seems like if `toCopyDefault` is True and `theme` is defined, then it would copy the theme's file instead of the default file. 

What do you think of the following flow:
If `theme` is undefined, then we will assume that the caller wants to copy the default bootstrap theme.
Else check if `theme` is a supported theme and copy over the relevant theme file.

jonahtanjz replied:
> What do you think of the following flow:
> If theme is undefined, then we will assume that the caller wants to copy the default bootstrap theme.
> Else check if theme is a supported theme and copy over the relevant theme file.

Hmm I think this will make a redundant copy of the default theme if no theme is specified in the initial build. Maybe I can rename `toCopyDefault` to `isRebuild` so that it is more concise. What do you think?

raysonkoh replied:
> Hmm I think this will make a redundant copy of the default theme if no theme is specified in the initial build. 

Hmm, I see. In that case, should we log a warning / throw exception for the case `toCopyDefault = True && theme`?

> Maybe I can rename toCopyDefault to isRebuild so that it is more concise. What do you think?

I am fine with either name.

jonahtanjz replied:
Sorry I think I wasn't too clear about this. I think I should rename `toCopyDefault` as it is a little misleading. This parameter should only be True if it is a rebuild (site.json was modified).

What do you think about this check:
```
if ((!isRebuild && !theme) || (theme && !_.has(SUPPORTED_THEMES_PATHS, theme))) {
      return _.noop;
}
```
It will be a `noop` during the first build if theme is not specified/theme is invalid. Else, if it is subsequent rebuild/theme is specified and valid, then, it will copy over the relevant theme's file (including default theme if `!theme`).

raysonkoh replied:
Yes that looks good! Can you also add a comment next to this check to indicate its purpose for future reference?

jonahtanjz replied:
Alright I've updated this and also added the comment :)

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -844,9 +844,39 @@ class Site {
+        || !_.isEqual(oldSiteConfig.style, this.siteConfig.style)
+        || !_.isEqual(oldSiteConfig.externalScripts, this.siteConfig.externalScripts)
+        || !_.isEqual(oldSiteConfig.globalOverride, this.siteConfig.globalOverride)
+        || !_.isEqual(oldSiteConfig.ignore, this.siteConfig.ignore)
```
👀 what's this for?

jonahtanjz replied:
This is to check if any of these configs have changed and if they have changed, would require a rebuild of all pages.

ang-zeyu replied:
(`ignore` in particular) Shouldn't be necessary I think; `ignore` is only used down the line for link validation warnings

jonahtanjz replied:
Ahh ok. I was under the impression that if a page is dependent on an asset that is later added to `ignore`, then that entire page will have to be rebuilt. Might have misunderstood the functionality of `ignore` here 😅

ang-zeyu reviewed (commented):
Thanks! @jonahtanjz 

> Since most of the attributes in site.json (except baseUrl, pages and deploy) will affect majority of the pages, hence all pages will be regenerated when any of these attributes (except those listed above) is modified.

Hmm shouldn't `baseUrl` be included? (`{{ baseUrl }}` variable changes).
Although, there might not be much value in supporting this in `serve` mode (since `baseUrl` would really only change for deployment), we can document this exception if so. 

As there are a few distinct groups / individual properties in this PR with separate logic some commit structure would be great here, e.g.:
- Support ignore...
- Support bootstrap theme...
- Support plugins... this needs some special handling as well (see `PluginManager`)
- Support "miscellaneous" page properties (all else under your `isGlobalConfigModified` function)
- ... another other appropriate ones

Separate PRs for each of these would be welcome instead as well

---

(optional) You can also consider implementing the fourth part of #1495 first as well to ensure all pages have the same updated copy of `SiteConfig`'s properties

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -889,6 +919,28 @@ class Site {
+   * Handles the reloading of ignore attributes
+   */
+  async handleIgnoreReload(oldIgnore) {
+    const assetsToRemove = _.difference(this.siteConfig.ignore, oldIgnore);
```
:eyes: should this be the other way round?

jonahtanjz replied:
Hmm I think this is right as new items that are added to ignore will have to be removed?

ang-zeyu replied:
👍 my bad, was looking from perspective of `siteConfig."include"`

jonahtanjz commented:
> Hmm shouldn't baseUrl be included? ({{ baseUrl }} variable changes).
> Although, there might not be much value in supporting this in serve mode (since baseUrl would really only change for deployment), we can document this exception if so.

Since {{ baseUrl }} is a variable change then I supposed it would make sense to be included so that the behaviour would be consistent.

> Plugins would need some special handling as well like themes and ignore. (see PluginManager)

For this portion were you referring to running `collectBaseUrl()` method? Since this is the only part where a new `PluginManager` is created and then subsequently assigned to all the new pages that will be rebuilt.

ang-zeyu reviewed (commented):
> Since {{ baseUrl }} is a variable change then I supposed it would make sense to be included so that the behaviour would be consistent.

@jonahtanjz its common and ok to forgo supporting something if the use case is practically zero, especially since `baseUrl` needs extra special care.

I might be missing a use case though. Could get a user opinion from @damithc here. (would live-reload for `baseUrl` parameter in site config be useful?)

> For this portion were you referring to running collectBaseUrl() method? Since this is the only part where a new PluginManager is created and then subsequently assigned to all the new pages that will be rebuilt.

Was thinking a more targeted approach that dosen't rerun the baseUrl collection. (can do a simple refactor / extraction of method here)
On that note (good catch), `External/LayoutManager`'s copies in `collectBaseUrl()` should be refreshed as well.

damithc commented:
> I might be missing a use case though. Could get a user opinion from @damithc here. (would live-reload for `baseUrl` parameter in site config be useful?)

I think it is OK not support live preview support for changes to the baseUrl as it is a very rare case. Just need to specify it clearly.

jonahtanjz commented:
@ang-zeyu @raysonkoh I've updated the PR to include the exception of not supporting live reload for `baseUrl` in the UG and refresh the `managers` when necessary. Let me know if I've missed out anything else :)

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -889,6 +927,42 @@ class Site {
+      || !_.isEqual(oldSiteConfig.plugins, this.siteConfig.plugins)
+      || !_.isEqual(oldSiteConfig.pluginsContext, this.siteConfig.pluginsContext)
+      || !_.isEqual(oldSiteConfig.disableHtmlBeautify, this.siteConfig.disableHtmlBeautify)
+      || !_.isEqual(oldSiteConfig.intrasiteLinkValidation, this.siteConfig.intrasiteLinkValidation)) {
```
missing `timeZone / locale / style` as well - should / could we just remove the checks here as well?
I feel there's not much value in further narrowing it down since `buildManagers` isn't very costly

jonahtanjz replied:
> I feel there's not much value in further narrowing it down since buildManagers isn't very costly

Alright since there's not much value for the checks as it is not costly, then `buildManagers` will be called for every full rebuild and whenever `ignore` is modified. I've updated the PR accordingly.

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -844,9 +851,39 @@ class Site {
+  }
+
+  /**
+   * Checks if any of the global attributes of site.json is modified
```
```suggestion
   * Checks if any attributes of site.json requiring a global rebuild are modified
```
`global attributes` might be a little vague

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -844,9 +851,39 @@ class Site {
     await this.readSiteConfig();
+    await this.handleIgnoreReload(oldSiteConfig);
+    await this.handlePageReload(oldAddressablePages, oldPagesSrc,
+                                oldSiteConfig, this.isGlobalConfigModified(oldSiteConfig));
```
hmm should we `isGlobalConfigModified` inside `handlePageReload`? since its not used anywhere else and concerns page rebuilding only

ang-zeyu replied:
sorry, actually just meant changing the call site. But the current changes are ok as well : )

jonahtanjz commented:
@ang-zeyu Thanks for the catch and suggestions. I've updated the PR to remove the further checks for `buildManagers` and also moved the `isGlobalConfigModified` method into `handlePageReload`.

ang-zeyu commented on docs/userGuide/siteJsonFile.md:
```diff
@@ -81,6 +81,10 @@ Here is a typical `site.json` file:
 
+<box type="warning">
+
+Note: `baseUrl` does not support live reload as it should not change in `markbind serve` mode.
```
Let's update the definition here as well with this https://deploy-preview-1529--markbind-master.netlify.app/userguide/glossary

Minor wording nit:
```suggestion
Note: `baseUrl` does not support live preview as there is no use case for changing it in during `markbind serve`.
```

Otherwise, Lgtm! 👍 

Let's wait and see if @raysonkoh has anything to add

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -1223,15 +1279,21 @@ class Site {
     const { theme } = this.siteConfig;
-    if (!theme || !_.has(SUPPORTED_THEMES_PATHS, theme)) {
+
+    // If it is the initial build using the default theme or if the theme specified
```
Let's stick to `/* ... */` for multiline comments

ang-zeyu commented on packages/core/src/Site/index.js:
```diff
@@ -856,8 +873,24 @@ class Site {
+    if (isGlobalConfigModified() || !_.isEmpty(addedPages) || !_.isEmpty(removedPages)) {
       await this.removeAsset(removedPages);
+      this.buildManagers();
       await this._rebuildSourceFiles();
```
Additional remark (no action needed): 

>> EDIT: After looking around at the other handlers (addHandler and removeHandler), it seems all pages should be rebuilt if the added/removed file is a page/dependency of page.
>
>This has been existing behaviour for a long while already but can probably be made into an issue 🙂

Per #1514's comment, the number of dependent pages generated / regenerated can be improved later as well.

jonahtanjz replied:
Alright. I'll open an issue for this and the previous discussion.

jonahtanjz commented:
@ang-zeyu @raysonkoh I've updated the PR accordingly :)

ang-zeyu commented on docs/userGuide/glossary.md:
```diff
@@ -10,8 +10,6 @@
 **_Live preview_** is:
 - Regeneration of affected content upon any change to <tooltip content="`.md`, `.mbd`, `.mbdf`, `.njk` files ... anything your content depends on!">source files</tooltip>, then reloading the updated site in the Browser.
 
-- Regeneration will also occur upon any modification to attributes in `site.json` with the exception of [`baseUrl`](siteJsonFile.md#baseurl-no-live-preview).
```
👀

jonahtanjz replied:
I've re-added this portion

raysonkoh commented:
LGTM 👍
